Sources/AriesFramework/agent/MessageReceiver.swift

```diff
@@ -40,12 +40,57 @@ public class MessageReceiver {
         }
     }
 
+    func findConnection(decryptedMessage: DecryptedMessageContext, message: AgentMessage) async throws -> ConnectionRecord? {
+        var connection = try await findConnectionByMessageKeys(decryptedMessage: decryptedMessage)
+        if connection == nil {
+            connection = try await findConnectionByMessageThreadId(message: message)
+            if connection != nil {
+                // If a connection is found by the message thread id, then this message is
+                // a connection-less exchange and recipient is the oob inviter.
+                // To be able to respond to the sender, sender's information should be updated
+                // based on incomming message because the oob inviter created a fake connection.
+                updateConnectionTheirDidDoc(&connection!, senderKey: decryptedMessage.senderKey)
+            }
+        }
+        return connection
+    }
+
     func findConnectionByMessageKeys(decryptedMessage: DecryptedMessageContext) async throws -> ConnectionRecord? {
         let connectionRecord = try await agent.connectionService.findByKeys(senderKey: decryptedMessage.senderKey ?? "",
                                                                             recipientKey: decryptedMessage.recipientKey ?? "")
         return connectionRecord
     }
 
+    func findConnectionByMessageThreadId(message: AgentMessage) async throws -> ConnectionRecord? {
+        guard let threadId = message.thread?.threadId else {
+            return nil
+        }
+        guard let oobRecord = try await agent.outOfBandService.findByAttachmentThreadId(threadId) else {
+            return nil
+        }
+        let connection = try await agent.connectionService.findByInvitationKey(oobRecord.outOfBandInvitation.invitationKey()!)
+        return connection
+    }
+
+    func updateConnectionTheirDidDoc(_ connection: inout ConnectionRecord, senderKey: String?) {
+        guard let senderKey = senderKey else {
+            return
+        }
+        let service = DidCommService(
```

Comment:
It would be better to add a comment here:
// TODO: get this from the message's service decorator if exists.